src/porepy/models/solution_strategy.py

```diff
@@ -451,12 +460,24 @@ def after_nonlinear_convergence(self) -> None:
         self.save_data_time_step()
 
     def after_nonlinear_failure(self) -> None:
-        """Method to be called if the non-linear solver fails to converge."""
+        """Method to be called if the non-linear solver fails to converge.
+
+        Parameters:
+            iteration_counter: Number of nonlinear solver iterations before failure.
+                Used for time step adaptation.
+
+        """
         self.save_data_time_step()
-        if self._is_nonlinear_problem():
+        if not self._is_nonlinear_problem():
+            raise ValueError("Tried solving singular matrix for the linear problem.")
```

Comment:
```suggestion
            raise ValueError("Failed to solve linear system for the linear problem.")
```
A user may switch to an iterative solver with poor settings (but non-singular matrix) without overriding this method